# Project 1: Training a Simple Neural Network with GPU

## Introduction

In this project, you will create, train, and evaluate a simple neural network using both TensorFlow and PyTorch. The objective is to ensure you are comfortable with setting up a neural network and utilizing GPU acceleration for training. 

The code for this assignment will mirror that of Lab 1, but you will need to complete some items, and make some changes.  You will use the [Fashion MNIST dataset](https://github.com/zalandoresearch/fashion-mnist) for this project.

## Objectives

1. Set up TensorFlow and PyTorch environments.
2. Verify GPU availability.
3. Implement a simple neural network in TensorFlow and PyTorch.
4. Train and evaluate the models.
5. Answer assessment questions.

## Instructions

Follow the steps below to complete the project. Ensure that you use a GPU to train your models.

---

### Step 1: Set Up Your Environment

First, install the necessary libraries. Run the following cell to install TensorFlow and PyTorch. 


Provide snapshots from your environment showing:
1) You are using a virtual environment
2) You have installed `TensorFlow` and `PyTorch`

---

### Step 2: Verify GPU Availability
Check if TensorFlow and PyTorch can detect the GPU.

Run the following two code blocks and show the output.

#### TensorFlow GPU Check

In [2]:
import tensorflow as tf

print("TensorFlow version:", tf.__version__)
gpu_devices = tf.config.list_physical_devices('GPU')
if gpu_devices:
    print("GPU is available for TensorFlow!")
else:
    print("No GPU found for TensorFlow.")

TensorFlow version: 2.15.0
GPU is available for TensorFlow!


#### PyTorch GPU Check

In [3]:
import torch

print("PyTorch version:", torch.__version__)
if torch.cuda.is_available():
    print("GPU is available for PyTorch!")
elif torch.mps.is_available():
    print("MPS (Apple Silicon GPU support) is available for PyTorch!")
else:
    print("No GPU found for PyTorch.")

PyTorch version: 2.5.1+cu121
GPU is available for PyTorch!


---

### Step 3: Implement and Train a Simple Neural Network
#### TensorFlow Implementation
1. Load and preprocess the Fashion MNIST dataset.
2. Define the neural network model.  Departing from the example in class, use 2 hidden layers, each with 64 units and relu activation functions, densely connected.
3. Compile the model.
4. Train the model using the GPU.  Use a batch size of 64, and run 6 epochs.
5. Evaluate the model.

You need to complete and run the code. Show the complete output.


In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical

# Load and preprocess the MNIST dataset
fashion_mnist_data = tf.keras.datasets.fashion_mnist.load_data()
(x_train, y_train), (x_test, y_test) = fashion_mnist_data
x_train = x_train / 255.0
x_test = x_test / 255.0
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

# Define the model
model = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(64, activation='relu'),
    Dense(10, activation='softmax')
])

# Compile the model
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train the model
# with tf.device('/GPU:0'):
model.fit(x_train, y_train, epochs=5, batch_size=32, validation_split=0.2)

# Evaluate the model
loss, accuracy = model.evaluate(x_test, y_test)
print(f'Test accuracy: {accuracy:.4f}')

2025-06-03 22:47:08.434501: I external/local_xla/xla/stream_executor/cuda/cuda_executor.cc:901] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
2025-06-03 22:47:08.437684: I external/local_xla/xla/stream_executor/cuda/cuda_executor.cc:901] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
2025-06-03 22:47:08.439822: I external/local_xla/xla/stream_executor/cuda/cuda_executor.cc:901] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-

Epoch 1/5


2025-06-03 22:47:08.847762: W external/local_tsl/tsl/framework/cpu_allocator_impl.cc:83] Allocation of 150528000 exceeds 10% of free system memory.
2025-06-03 22:47:08.960419: W external/local_tsl/tsl/framework/cpu_allocator_impl.cc:83] Allocation of 150528000 exceeds 10% of free system memory.
2025-06-03 22:47:10.020873: I external/local_xla/xla/service/service.cc:168] XLA service 0x7f64cf93f050 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:
2025-06-03 22:47:10.020908: I external/local_xla/xla/service/service.cc:176]   StreamExecutor device (0): NVIDIA GeForce GTX 1660 Ti, Compute Capability 7.5
2025-06-03 22:47:10.028702: I tensorflow/compiler/mlir/tensorflow/utils/dump_mlir_util.cc:269] disabling MLIR crash reproducer, set env var `MLIR_CRASH_REPRODUCER_DIRECTORY` to enable.
2025-06-03 22:47:10.061966: I external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:454] Loaded cuDNN version 8905
I0000 00:00:1749005230.136164    9996 device_compiler

1500/1500 [==============================] - 5s 2ms/step - loss: 0.5308 - accuracy: 0.8151 - val_loss: 0.4239 - val_accuracy: 0.8493
Epoch 2/5
1500/1500 [==============================] - 3s 2ms/step - loss: 0.3867 - accuracy: 0.8608 - val_loss: 0.3763 - val_accuracy: 0.8620
Epoch 3/5
1500/1500 [==============================] - 3s 2ms/step - loss: 0.3481 - accuracy: 0.8712 - val_loss: 0.3666 - val_accuracy: 0.8669
Epoch 4/5
1500/1500 [==============================] - 3s 2ms/step - loss: 0.3206 - accuracy: 0.8819 - val_loss: 0.3387 - val_accuracy: 0.8762
Epoch 5/5
313/313 [==============================] - 0s 1ms/step - loss: 0.3670 - accuracy: 0.8690
Test accuracy: 0.8690


#### PyTorch Implementation
1. Load and preprocess the Fashion MNIST dataset.  Set batch size to 64.
2. Define the neural network model.  Departing from the example in class, use 2 hidden layers, each with 64 units and relu activation functions, densely connected.  
4. Define loss function and optimizer.
5. Train the model using the GPU.  Run 6 epochs.
6. Evaluate the model.

You need to complete and run the code. Show the complete output.


In [ ]:
import torch
import torch.nn as nn
import torch.optim as optim
from torchvision import datasets, transforms
from torch.utils.data import DataLoader

# Load and preprocess the MNIST dataset
transform = transforms.Compose([transforms.ToTensor(), transforms.Normalize((0.5,), (0.5,))])
train_dataset = datasets.FashionMNIST(root='./data', train=True, transform=transform, download=True)
test_dataset = datasets.FashionMNIST(root='./data', train=False, transform=transform)

train_loader = DataLoader(train_dataset, batch_size=64, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=64, shuffle=False)

# Define the model
class SimpleNN(nn.Module):
    def __init__(self):
        super(SimpleNN, self).__init__()
        self.flatten = nn.Flatten()
        self.fc1 = nn.Linear(28 * 28, 128)  # 28x28 input flattened → 128 neurons
        self.relu = nn.ReLU()
        self.fc2 = nn.Linear(128, 10)  
    
    def forward(self, x):
        x = self.flatten(x)
        x = self.relu(self.fc1(x))
        x = self.fc2(x)
        return x

model = SimpleNN()

# Define loss and optimizer
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=0.001)

# Check for GPU
# device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
# model.to(device)
# Check for GPU
if torch.cuda.is_available():
    dev = 'cuda'
elif torch.mps.is_available():
    dev = 'mps'
else:
    dev = 'cpu'
device = torch.device(dev)
model.to(device)


# Train the model
num_epochs = 6
for epoch in range(num_epochs):
    model.train()
    for images, labels in train_loader:
        images, labels = images.to(device), labels.to(device)
        
        # Forward pass
        outputs = model(images)
        loss = criterion(outputs, labels)
        
        # Backward and optimize
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
    
    print(f'Epoch [{epoch+1}/{num_epochs}], Loss: {loss.item():.4f}')

# Evaluate the model
model.eval()
with torch.no_grad():
    correct = 0
    total = 0
    for images, labels in test_loader:
        images, labels = images.to(device), labels.to(device)
        outputs = model(images)
        _, predicted = torch.max(outputs.data, 1)
        total += labels.size(0)
        correct += (predicted == labels).sum().item()
    
    accuracy = 100 * correct / total
    print(f'Test Accuracy: {accuracy:.2f}%')

100%|██████████| 26.4M/26.4M [00:02<00:00, 10.4MB/s]


Extracting ./data/FashionMNIST/raw/train-images-idx3-ubyte.gz to ./data/FashionMNIST/raw



100%|██████████| 29.5k/29.5k [00:00<00:00, 266kB/s]


Extracting ./data/FashionMNIST/raw/train-labels-idx1-ubyte.gz to ./data/FashionMNIST/raw



100%|██████████| 4.42M/4.42M [00:00<00:00, 4.94MB/s]


Extracting ./data/FashionMNIST/raw/t10k-images-idx3-ubyte.gz to ./data/FashionMNIST/raw



100%|██████████| 5.15k/5.15k [00:00<00:00, 4.24MB/s]


Extracting ./data/FashionMNIST/raw/t10k-labels-idx1-ubyte.gz to ./data/FashionMNIST/raw

Epoch [1/6], Loss: 0.5069
Epoch [2/6], Loss: 0.6438
Epoch [3/6], Loss: 0.3031
Epoch [4/6], Loss: 0.2903
Epoch [5/6], Loss: 0.1610
Epoch [6/6], Loss: 0.3480
Test Accuracy: 87.33%


---
### Questions
Answer the following questions in detail.

1. What is the purpose of normalizing the input data in both TensorFlow and PyTorch implementations?
2. Explain the role of the activation function relu in the neural network.
3. Why is it important to use GPU for training neural networks?
4. Compare the training time and accuracy of the TensorFlow and PyTorch models. Which one performed better and why?


1. Answer:
- The process of training neural networks becomes faster when inputs are on a similar level of scale.
- The effect of gradients is smoother and more helpful when data has been normalized.
- Shares out input values more evenly, so that no neuron is always recognized as the main one.
- It helps avoid problems when gradients become too small or too large.
- Better generalization and generally higher accuracy are often the result.
- Helps make the process of optimization more effective.

2. Answer:
In other words, it provides neural networks with the ability to use data that is not linearly arranged.
Very efficient to compute. Gradients are usually stronger in the positive range which prevents issues seen in other functions.
By giving output 0 for negative inputs, logistic regression helps the model be more efficient and reliable.

3. Answer:
Processing thousands of operations all at once is possible in GPUs, making them excellent for deep learning calculations.
GPUs are at least 10 times quicker than CPUs for most machine learning applications.
Looking at it from the perspective of resources, faster computing leads to saving time and energy on results, so people can quickly try out ideas and test them out.
Since deep learning models today have millions of parameters, training them would take days on a CPU, but with GPUs, this is achieved in hours.

4. ANswer:
 - Training Time
  TensorFlow trained significantly faster 10 sec compared to PyTorch 31 sec on this task. This is beacuse TensorFlow’s internal optimizations of various parameters.

  - Accuracy
  Both models achieved very similar accuracy , which is expected since the network architectures and training settings were nearly identical. PyTorch was slightly higher 87.11% than TensorFlow 86.90%, but this small difference is within the range of random variation for different training runs.


---
### Submission
Submit a link to your completed Jupyter Notebook (e.g., on GitHub (private) or Google Colab) with all the cells executed, and answers to the assessment questions included at the end of the notebook.